src/lib/core/selection/selection.ts

```diff
@@ -0,0 +1,130 @@
+import {Subject} from 'rxjs/Subject';
+
+
+/**
+ * Class to be used to power selecting one or more options from a list.
+ * @docs-private
+ */
+export class SelectionModel<T> {
+  /** Currently-selected values. */
+  private _selection: Set<T> = new Set();
+
+  /** Keeps track of the deselected options that haven't been emitted by the change event. */
+  private _deselectedToEmit: T[] = [];
+
+  /** Keeps track of the selected option that haven't been emitted by the change event. */
+  private _selectedToEmit: T[] = [];
+
+  /** Cache for the array value of the selected items. */
+  private _selected: T[];
+
+  /** Selected value(s). */
+  get selected(): T[] {
+    if (!this._selected) {
+      this._selected = Array.from(this._selection.values());
+    }
+
+    return this._selected;
+  }
+
+  /** Event emitted when the value has changed. */
+  onChange: Subject<SelectionChange<T>> = new Subject();
+
+  constructor(private _isMulti = false, initiallySelectedValues?: T[]) {
+    if (initiallySelectedValues) {
+      if (_isMulti) {
+        initiallySelectedValues.forEach(value => this._select(value));
+      } else {
+        this._select(initiallySelectedValues[0]);
+      }
+
+      // Clear the array in order to avoid firing the change event for preselected values.
+      this._selectedToEmit.length = 0;
+    }
+  }
+
+  /**
+   * Selects a value or an array of values.
+   */
+  select(value: T): void {
```

Comment:
`_markSelected` makes sense, but should we rename `_deselect` and `_clear` in that case as well?